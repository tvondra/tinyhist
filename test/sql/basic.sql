\set ECHO none

-- disable the notices for the create script (shell types etc.)
SET client_min_messages = 'WARNING';
\i sql/tinyhist--1.0.0.sql
SET client_min_messages = 'NOTICE';

\set ECHO all

/* simple case */
SELECT tinyhist_agg(i) FROM generate_series(1,10000) s(i);

SELECT * FROM tinyhist_info((SELECT tinyhist_agg(i) FROM generate_series(1,10000) s(i)));

SELECT * FROM tinyhist_buckets((SELECT tinyhist_agg(i) FROM generate_series(1,10000) s(i)));

SELECT (SELECT tinyhist_agg(i) FROM generate_series(1,5000) s(i)) + (SELECT tinyhist_agg(i) FROM generate_series(5001,10000) s(i));

/* wider range */
SELECT tinyhist_agg(i*10) FROM generate_series(1,10000) s(i);

SELECT * FROM tinyhist_info((SELECT tinyhist_agg(i*10) FROM generate_series(1,10000) s(i)));

SELECT * FROM tinyhist_buckets((SELECT tinyhist_agg(i*10) FROM generate_series(1,10000) s(i)));

SELECT (SELECT tinyhist_agg(i * 10) FROM generate_series(1,5000) s(i)) + (SELECT tinyhist_agg(i * 10) FROM generate_series(5001,10000) s(i));

/* aggregate pre-aggregated histograms */
CREATE TABLE tinyhist_test (h tinyhist);

INSERT INTO tinyhist_test SELECT tinyhist_agg(i) FROM generate_series(1,10000) s(i) GROUP BY (random() * 10)::int;
SELECT tinyhist_agg(h) FROM tinyhist_test;

TRUNCATE TABLE tinyhist_test;

INSERT INTO tinyhist_test SELECT tinyhist_agg(i * 10) FROM generate_series(1,10000) s(i) GROUP BY (random() * 10)::int;
SELECT tinyhist_agg(h) FROM tinyhist_test;

DROP TABLE tinyhist_test;
