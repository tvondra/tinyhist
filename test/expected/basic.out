\set ECHO none
/* simple case */
SELECT tinyhist_agg(i) FROM generate_series(1,10000) s(i);
                                tinyhist_agg                                 
-----------------------------------------------------------------------------
 {0, 0, 1, 1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 1808, 0}
(1 row)

SELECT * FROM tinyhist_info((SELECT tinyhist_agg(i) FROM generate_series(1,10000) s(i)));
 hist_unit | hist_sample_rate | hist_count | hist_upper 
-----------+------------------+------------+------------
         1 |                1 |      10000 |      32768
(1 row)

SELECT * FROM tinyhist_buckets((SELECT tinyhist_agg(i) FROM generate_series(1,10000) s(i)));
 bucket_index | bucket_lower | bucket_upper | bucket_range | bucket_count | bucket_frac | bucket_density 
--------------+--------------+--------------+--------------+--------------+-------------+----------------
            0 |            0 |            1 |            1 |            1 |      0.0001 |         0.0001
            1 |            1 |            2 |            1 |            1 |      0.0001 |         0.0001
            2 |            2 |            4 |            2 |            2 |      0.0002 |         0.0001
            3 |            4 |            8 |            4 |            4 |      0.0004 |         0.0001
            4 |            8 |           16 |            8 |            8 |      0.0008 |         0.0001
            5 |           16 |           32 |           16 |           16 |      0.0016 |         0.0001
            6 |           32 |           64 |           32 |           32 |      0.0032 |         0.0001
            7 |           64 |          128 |           64 |           64 |      0.0064 |         0.0001
            8 |          128 |          256 |          128 |          128 |      0.0128 |         0.0001
            9 |          256 |          512 |          256 |          256 |      0.0256 |         0.0001
           10 |          512 |         1024 |          512 |          512 |      0.0512 |         0.0001
           11 |         1024 |         2048 |         1024 |         1024 |      0.1024 |         0.0001
           12 |         2048 |         4096 |         2048 |         2048 |      0.2048 |         0.0001
           13 |         4096 |         8192 |         4096 |         4096 |      0.4096 |         0.0001
           14 |         8192 |        16384 |         8192 |         1808 |      0.1808 | 2.20703125e-05
           15 |        16384 |        32768 |        16384 |            0 |           0 |              0
(16 rows)

/* wider range */
SELECT tinyhist_agg(i*10) FROM generate_series(1,10000) s(i);
                                tinyhist_agg                                
----------------------------------------------------------------------------
 {0, 2, 0, 0, 1, 2, 3, 6, 13, 26, 51, 102, 205, 410, 819, 1638, 3277, 3447}
(1 row)

SELECT * FROM tinyhist_info((SELECT tinyhist_agg(i*10) FROM generate_series(1,10000) s(i)));
 hist_unit | hist_sample_rate | hist_count | hist_upper 
-----------+------------------+------------+------------
         4 |                1 |      10000 |     131072
(1 row)

SELECT * FROM tinyhist_buckets((SELECT tinyhist_agg(i*10) FROM generate_series(1,10000) s(i)));
 bucket_index | bucket_lower | bucket_upper | bucket_range | bucket_count | bucket_frac |   bucket_density    
--------------+--------------+--------------+--------------+--------------+-------------+---------------------
            0 |            0 |            4 |            4 |            0 |           0 |                   0
            1 |            4 |            8 |            4 |            0 |           0 |                   0
            2 |            8 |           16 |            8 |            1 |      0.0001 |               5e-05
            3 |           16 |           32 |           16 |            2 |      0.0002 |               5e-05
            4 |           32 |           64 |           32 |            3 |      0.0003 |            3.75e-05
            5 |           64 |          128 |           64 |            6 |      0.0006 |            3.75e-05
            6 |          128 |          256 |          128 |           13 |      0.0013 |          4.0625e-05
            7 |          256 |          512 |          256 |           26 |      0.0026 |          4.0625e-05
            8 |          512 |         1024 |          512 |           51 |      0.0051 |        3.984375e-05
            9 |         1024 |         2048 |         1024 |          102 |      0.0102 |        3.984375e-05
           10 |         2048 |         4096 |         2048 |          205 |      0.0205 |      4.00390625e-05
           11 |         4096 |         8192 |         4096 |          410 |       0.041 |      4.00390625e-05
           12 |         8192 |        16384 |         8192 |          819 |      0.0819 |    3.9990234375e-05
           13 |        16384 |        32768 |        16384 |         1638 |      0.1638 |    3.9990234375e-05
           14 |        32768 |        65536 |        32768 |         3277 |      0.3277 |  4.000244140625e-05
           15 |        65536 |       131072 |        65536 |         3447 |      0.3447 | 2.1038818359375e-05
(16 rows)

